File: 62-unique-paths/62-unique-paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        return self.solve(m, n, 0, 0, {})
        
    def solve(self, m: int, n: int, r: int, c: int, dp: List) -> int:
        if (r, c) in dp:
            return dp[(r, c)]
        
        if r == m - 1 and c == n - 1:
            return 1
        
        if r >= m or c >= n:
            return 0
        
        down = self.solve(m, n, r + 1, c, dp)
        right = self.solve(m, n, r, c + 1, dp)
        
        dp[(r, c)] = down + right
        return dp[(r, c)]

File: 62-unique-paths/62-unique-paths.cpp
class Solution {
public:
    int uniquePaths(int m, int n) {
        map<pair<int, int>, int> dp;
        return solve(m - 1, n - 1, 0, 0, dp);
    }
    
    int solve(int m, int n, int r, int c, map<pair<int, int>, int> &dp) {
        if(dp.find({r, c}) != dp.end()) return dp[{r, c}];
        
        if((r == m) && (c == n)) return 1;
        
        if((r > m) || (c > n)) return 0;
        
        dp[{r, c}] = solve(m, n, r + 1, c, dp) + solve(m, n, r, c + 1, dp);
        
        return dp[{r, c}];
    }
};